pkg/util/pki/csr.go

```diff
@@ -303,37 +328,67 @@ func GenerateTemplate(crt *v1.Certificate) (*x509.Certificate, error) {
 		return nil, err
 	}
 
-	return &x509.Certificate{
-		// Version must be 2 according to RFC5280.
-		// A version value of 2 confusingly means version 3.
-		// This value isn't used by Go at the time of writing.
-		// https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.1
-		Version:               2,
-		BasicConstraintsValid: true,
-		SerialNumber:          serialNumber,
-		PublicKeyAlgorithm:    pubKeyAlgo,
-		IsCA:                  crt.Spec.IsCA,
-		Subject: pkix.Name{
-			Country:            subject.Countries,
-			Organization:       organization,
-			OrganizationalUnit: subject.OrganizationalUnits,
-			Locality:           subject.Localities,
-			Province:           subject.Provinces,
-			StreetAddress:      subject.StreetAddresses,
-			PostalCode:         subject.PostalCodes,
-			SerialNumber:       subject.SerialNumber,
-			CommonName:         commonName,
-		},
-		NotBefore: time.Now(),
-		NotAfter:  time.Now().Add(certDuration),
-		// see http://golang.org/pkg/crypto/x509/#KeyUsage
-		KeyUsage:       keyUsages,
-		ExtKeyUsage:    extKeyUsages,
-		DNSNames:       dnsNames,
-		IPAddresses:    ipAddresses,
-		URIs:           uris,
-		EmailAddresses: crt.Spec.EmailAddresses,
-	}, nil
+	if len(crt.Spec.LiteralSubject) > 0 { // todo: feature gate?
+		rawSubject, err := ParseSubjectStringToRawDerBytes(crt.Spec.LiteralSubject)
```

Comment:
Yes, makes sense!

Reply:
On the other hand. What exactly do we check there? The only added call there is to the same parse function plus a check for the common name which is what I added now.

Reply:
I was thinking a check for keys like the one that you already added in validating webhook https://github.com/spockz/cert-manager/blob/add-literal-subject-field-to-certificate/internal/apis/certmanager/validation/certificate.go#L76-L83

Theoretically an invalid key should not pass the webhook validation of course, perhaps I am being too paranoid- but feeling a little concerned with passing these untyped keys

Reply:
Okay, I can understand that. But that raises the question on the implementation. Just call the same validation method because why copy the implementation? Then, if we are doing the validation in the controller again anyway why do we have a webhook doing validation? (Maybe because of the API of the validating webhook it can more easily give nice error messages. I'm not *that* deep into the K8s ecosystem.)